mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/server/master/scheduler/ResourceClusterAwareSchedulerActor.java

```diff
@@ -0,0 +1,318 @@
+    }
+
+    public ResourceClusterAwareSchedulerActor(
+        int maxScheduleRetries,
+        int maxCancelRetries,
+        Duration intervalBetweenRetries,
+        ResourceCluster resourceCluster,
+        ExecuteStageRequestFactory executeStageRequestFactory,
+        JobMessageRouter jobMessageRouter) {
+        this.resourceCluster = resourceCluster;
+        this.executeStageRequestFactory = executeStageRequestFactory;
+        this.jobMessageRouter = jobMessageRouter;
+        this.maxScheduleRetries = maxScheduleRetries;
+        this.intervalBetweenRetries = intervalBetweenRetries;
+        this.maxCancelRetries = maxCancelRetries;
+    }
+
+    @Override
+    public Receive createReceive() {
+        return ReceiveBuilder.create()
+            .match(ScheduleRequestEvent.class, this::onScheduleRequestEvent)
+            .match(InitializeRunningWorkerRequestEvent.class, this::onInitializeRunningWorkerRequest)
+            .match(CancelRequestEvent.class, this::onCancelRequestEvent)
+            .match(AssignedScheduleRequestEvent.class, this::onAssignedScheduleRequestEvent)
+            .match(FailedToScheduleRequestEvent.class, this::onFailedScheduleRequestEvent)
+            .match(SubmittedScheduleRequestEvent.class, this::onSubmittedScheduleRequestEvent)
+            .match(FailedToSubmitScheduleRequestEvent.class, this::onFailedToSubmitScheduleRequestEvent)
+            .match(RetryCancelRequestEvent.class, this::onRetryCancelRequestEvent)
+            .match(Noop.class, this::onNoop)
+            .build();
+    }
+
+    private void onScheduleRequestEvent(ScheduleRequestEvent event) {
+        if (event.isRetry()) {
+            log.info("Retrying Schedule Request {}, attempt {}", event.getScheduleRequest(),
+                event.getAttempt());
+        }
+
+        CompletableFuture<Object> assignedFuture =
+            resourceCluster
+                .getTaskExecutorFor(event.scheduleRequest.getMachineDefinition(),
+                    event.scheduleRequest.getWorkerId())
+                .<Object>thenApply(
+                    taskExecutorID1 -> new AssignedScheduleRequestEvent(event.getScheduleRequest(),
+                        taskExecutorID1))
+                .exceptionally(event::onFailure);
+
+        pipe(assignedFuture, getContext().getDispatcher()).to(self());
+    }
+
+    private void onInitializeRunningWorkerRequest(InitializeRunningWorkerRequestEvent request) {
+        resourceCluster.initializeTaskExecutor(
+            request.getTaskExecutorID(),
+            request.getScheduleRequest().getWorkerId());
+    }
+
+    private void onAssignedScheduleRequestEvent(AssignedScheduleRequestEvent event) {
+        try {
+            TaskExecutorGateway gateway =
+                resourceCluster.getTaskExecutorGateway(event.getTaskExecutorID()).join();
+
+            TaskExecutorRegistration info =
+                resourceCluster.getTaskExecutorInfo(event.getTaskExecutorID()).join();
```

Comment:
Are the two 'get' calls here involve any long IO ops for blocking concern?

Reply:
No, these should be pretty cheap. The only reason they get async responses is because they send fire-and-forget messages to the actors involved.